File: packages/react-native-nitro-web-image/ios/HybridImageView.swift
//
//  HybridImageView.swift
//  react-native-nitro-web-image
//
//  Created by Marc Rousavy on 10.06.25.
//

import Foundation
import UIKit

class HybridImageView: HybridNitroImageViewSpec {
  let imageView = UIImageView(image: nil)
  var view: UIView { imageView }

  var resizeMode: ResizeMode? {
    didSet {
      let mode = resizeMode ?? .contain
      switch mode {
        case .cover:
            imageView.contentMode = .scaleAspectFill
            imageView.clipsToBounds = true
        case .contain:
            imageView.contentMode = .scaleAspectFit
        case .stretch:
            imageView.contentMode = .scaleToFill
        case .center:
            imageView.contentMode = .center
      }
      updateImage()
    }
  }

  var image: (any HybridImageSpec)? {
    didSet {
      updateImage()
    }
  }

  func updateImage() {
    guard let hybridImage = image as? HybridImage else { return }
    print("Updating ImageView's image...")

    print("Frame: \(imageView.frame)")
    imageView.image = hybridImage.uiImage
  }
}

File: packages/react-native-nitro-web-image/android/src/main/java/com/margelo/nitro/web/image/HybridImageView.kt
package com.margelo.nitro.web.image

import android.content.Context
import android.view.View
import android.widget.ImageView

class HybridImageView(context: Context): HybridNitroImageViewSpec() {
    override var resizeMode: ResizeMode? = ResizeMode.CONTAIN
        set(value) {
            field = value
            imageView.scaleType = when (value) {
                ResizeMode.COVER -> ImageView.ScaleType.CENTER_CROP
                ResizeMode.CONTAIN -> ImageView.ScaleType.FIT_CENTER
                ResizeMode.STRETCH -> ImageView.ScaleType.FIT_XY
                ResizeMode.CENTER -> ImageView.ScaleType.CENTER
                null -> ImageView.ScaleType.FIT_CENTER
            }
            updateImage()
        }

    override var image: HybridImageSpec? = null
        set(value) {
            field = value
            updateImage()
        }

    val imageView = ImageView(context)
    override val view: View = imageView

    private fun updateImage() {
        val image = image as? HybridImage ?: return
        imageView.setImageBitmap(image.bitmap)
    }
}
